packages/fxa-auth-server/lib/payments/stripe.ts

```diff
@@ -745,6 +745,88 @@ export class StripeHelper {
     return promotionCode;
   }
 
+  /**
+   * Verify whether either a Promotion Code or Coupon is valid.
+   * If the properties are invalid, then return what's making it invalid.
+   * Current invalid states include:
+   * - Expired
+   * - Maximally redeemed
+   */
+  verifyPromotionAndCouponProperties({
+    valid,
+    redeem_by: redeemBy,
+    max_redemptions: maxRedemptions,
+    times_redeemed: timesRedeemed,
+  }: {
+    valid: boolean;
+    redeem_by: number | null;
+    max_redemptions: number | null;
+    times_redeemed: number;
+  }) {
+    let expired = false;
+    let maximallyRedeemed = false;
+
+    if (!valid) {
+      if (redeemBy) {
+        const expiry = new Date(redeemBy * 1000);
+        const now = new Date();
+        expired = now > expiry;
+      } else {
+        expired = false;
+      }
+
+      if (maxRedemptions) {
+        maximallyRedeemed = timesRedeemed >= maxRedemptions;
+      } else {
+        maximallyRedeemed = false;
+      }
+    }
+
+    return {
+      valid,
+      expired,
+      maximallyRedeemed,
+    };
+  }
+
+  /**
+   * Verify that the Promotion Code and Coupon are valid.
+   * If either are not valid, then return what's making it invalid.
+   * Current invalid states include:
+   * - Expired
+   * - Maximally redeemed
+   */
+  verifyPromotionAndCoupon(promotionCode: Stripe.PromotionCode) {
+    const { coupon } = promotionCode;
+    const returnValue: {
+      valid: boolean;
+      expired?: boolean;
+      maximallyRedeemed?: boolean;
+    } = { valid: false };
+
+    const verifyCoupon = this.verifyPromotionAndCouponProperties(coupon);
+    const verifyPromotionCode = this.verifyPromotionAndCouponProperties({
+      ...promotionCode,
+      valid: promotionCode.active,
+      redeem_by: promotionCode.expires_at,
+    });
+
+    returnValue.valid = verifyCoupon.valid && verifyPromotionCode.valid;
+
+    if (verifyCoupon.expired || verifyPromotionCode.expired) {
+      returnValue.expired = true;
+    }
+
+    if (
+      verifyCoupon.maximallyRedeemed ||
+      verifyPromotionCode.maximallyRedeemed
+    ) {
+      returnValue.maximallyRedeemed = true;
+    }
+
+    return returnValue;
```

Comment:
Maybe just...
```typescript
return {
  expired: verifyCoupon.expired || verifyPromotionCode.expired,
  valid: verifyCoupon.valid && verifyPromotionCode.valid,
  maximallyRedeemed: verifyCoupon.maximallyRedeemed || verifyPromotionCode.maximallyRedeemed
}
```

Then you could also drop its initialization.

Reply:
I'm also not super happy with my current implementation, but I mainly did it this way to keep the API behavior the same.

Currently we only return `expired` and `maximallyRedeemed` when they have a value of `true`, else the properties aren't included in the return object.

Is there a better way to achieve this, or should we change the logic to return true/false instread of true/undefined.

Reply:
Is there any harm in always returning them even when they're false?

Reply:
Nope, I don't think so. Just changing a bunch of tests, but that's about it.

Reply:
I'm fine leaving it or changing it, if its a whole bunch of tests, maybe we file a bug to trim it down a bit later with that simplification.